src/basic/path-util.c

```diff
@@ -637,6 +637,9 @@ int find_binary(const char *name, char **ret) {
                 if (!j)
                         return -ENOMEM;
 
+                if (is_dir(j, true))
```

Comment:
this check sucks for two reasons:

error handling is borked. i.e. is_dir() might return < 0 if something is off (for example no access to some dir earlier in the path), and you treat this as if it was a dir, and don't update last_error.

it's a second syscall, which is ugly in particular as you invoke this in pretty much every iteration, as it is unikely to ever trigger except in niche cases.

Hence, please do it like this:

```c
if (access(j, X_OK) >= 0) { /* Does it have the x bit set? */
        _cleanup_free_ char *with_dash;

        with_dash = strjoin(j, "/");
        if (!with_dash)
                return -ENOMEM;

        if (access(with_dash, X_OK) >= 0) /* Check again, this time with dash appended */
                continue; /* It's a directory, not interesting to us then */
        if (errno == ENOTDIR) {
                 /* Found it! */
                …
```

This makes use of the fact that access() checks if something is a dir if the path is suffixed with "/".

This has two benefits: in the typical case where the binary is not found in the first $PATH element we'll just do one syscall instead of two. And we use the cheaper access() thant stat(). And we check for all errors

Reply:
mmm, yes. another one incoming.

Reply:
#16857 is out. Thanks!